### PR TITLE
Add a11y attributes and navigation for `date`, `checkbox` and `handsontable` types.

### PR DESCRIPTION
**Pull request by jansiegel**
### Context
This PR:
- Adds the aria tags described in handsontable/dev-handsontable#1440.
- Implements the navigation suggestions from handsontable/dev-handsontable#1440.
- Corrects some things that were missing after finishing #10529
- Implements `handsontable` and `date` renderers to make adding aria tags cleaner.

### How has this been tested?
Tested manually + added test cases.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature or improvement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Additional language file or change to the existing one (translations)

### Related issue(s):
1. handsontable/dev-handsontable#1440

### Affected project(s):
- [x] `handsontable`
- [ ] `@handsontable/angular`
- [ ] `@handsontable/react`
- [ ] `@handsontable/vue`
- [ ] `@handsontable/vue3`

### Checklist:

- [x] I have reviewed the guidelines about [Contributing to Handsontable](https://github.com/handsontable/handsontable/blob/master/CONTRIBUTING.md) and I confirm that my code follows the code style of this project.
- [x] I have signed the [Contributor License Agreement](https://docs.google.com/forms/d/e/1FAIpQLScpMq4swMelvw3-onxC8Jl29m0fVp5hpf7d1yQVklqVjGjWGA/viewform?c=0&w=1)
- [ ] My change requires a change to the documentation.

### CODE REVIEW

File: handsontable/src/editors/dateEditor/dateEditor.js [hunk 1]
```diff
@@ -138,10 +151,16 @@ export class DateEditor extends TextEditor {
       this.$datePicker.destroy();
     }
 
-    this.instance._registerTimeout(() => {
-      this.instance._refreshBorders();
+    this.hot._registerTimeout(() => {
+      this.hot._refreshBorders();
     });
 
+    if (this.TD && this.hot.getSettings().ariaTags) {
```
Comment (budnix) [inline handsontable/src/editors/dateEditor/dateEditor.js:158]:
```suggestion
    if (this.hot.getSettings().ariaTags) {
```
Isn't the TD element always available?

**Review by budnix — CHANGES_REQUESTED** [review]
It'd be nice to prevent the cursor movement in the editor. The issue happens when I navigating the dates.
Kapture 2023-10-13 at 10 02 55

File: handsontable/src/editors/handsontableEditor/__tests__/a11y/navigation.spec.js [hunk 2]
```diff
@@ -0,0 +1,122 @@
+describe('a11y navigation', () => {
```
Comment (budnix) [inline handsontable/src/editors/handsontableEditor/__tests__/a11y/navigation.spec.js:1]: The same as https://github.com/handsontable/handsontable/pull/10535#discussion_r1357907159.

**Comment by budnix** [comment]
I don't know if it's out of the scope of this PR, but it's impossible to navigate through the Handsontable cell-type editor. Only vertical navigation works.
Kapture 2023-10-13 at 10 07 37

File: handsontable/src/editors/handsontableEditor/handsontableEditor.js [hunk 3]
```diff
@@ -46,6 +47,12 @@ export class HandsontableEditor extends TextEditor {
 
     setCaretPosition(this.TEXTAREA, 0, this.TEXTAREA.value.length);
     this.refreshDimensions();
+
+    if (this.hot.getSettings().ariaTags) {
+      setAttribute(this.TD, [
+        A11Y_EXPANDED('true'),
```
Comment (budnix) [inline handsontable/src/editors/handsontableEditor/handsontableEditor.js:53]: There is a missing test for that ARIA tag.

**Comment by jansiegel** [comment]
> I don't know if it's out of the scope of this PR, but it's impossible to navigate through the Handsontable cell-type editor. Only vertical navigation works. Kapture 2023-10-13 at 10 07 37 [ Kapture 2023-10-13 at 10 07 37 ](https://user-images.githubusercontent.com/571316/274838079-6bf2eb83-b294-4dbc-87bb-3cf6dc79f283.gif) [ ](https://user-images.githubusercontent.com/571316/274838079-6bf2eb83-b294-4dbc-87bb-3cf6dc79f283.gif)

I think it works the same way in the current version; is that correct?
https://handsontable.com/docs/javascript-data-grid/handsontable-cell-type/

**Comment by budnix** [comment]
> I think it works the same way in the current version; is that correct? https://handsontable.com/docs/javascript-data-grid/handsontable-cell-type/

Yes, but since we improve all cell types it sounds natural to fix that too.

**Comment by jansiegel** [comment]
> > I think it works the same way in the current version; is that correct? https://handsontable.com/docs/javascript-data-grid/handsontable-cell-type/
> 
> Yes, but since we improve all cell types it sounds natural to fix that too.

I guess it's a question to @evanSe, should we extend the scope of this task with fixing the problem @budnix described?

**Comment by evanSe** [comment]
> > > I think it works the same way in the current version; is that correct? https://handsontable.com/docs/javascript-data-grid/handsontable-cell-type/
> > 
> > 
> > Yes, but since we improve all cell types it sounds natural to fix that too.
> 
> I guess it's a question to @evanSe, should we extend the scope of this task with fixing the problem @budnix described?

@jansiegel  we should improve this in the future, purely targeting whats in scope for the ticket for now.

File: handsontable/src/editors/dateEditor/dateEditor.js (hunk 1 above)
Comment (budnix) [reply handsontable/src/editors/dateEditor/dateEditor.js:158]: Thanks for explanation 👍

File: handsontable/src/editors/dateEditor/dateEditor.js [hunk 4]
```diff
@@ -72,7 +73,22 @@ export class DateEditor extends TextEditor {
...
+      }
+
+      event.stopPropagation();
+    });
+
+    /**
+     * Prevent caret movement in the TEXTAREA when navigating over the date picker.
+     */
+    eventManager.addEventListener(this.TEXTAREA, 'keydown', (event) => {
+      if (['ArrowUp', 'ArrowDown', 'ArrowLeft', 'ArrowRight'].includes(event.code)) {
```
Comment (budnix) [inline handsontable/src/editors/dateEditor/dateEditor.js:88]: Can I ask to cover that fix with caret movement with tests?

**Review by budnix — CHANGES_REQUESTED** [review]
Everything looks good 👌 except for one missing test that would be worth adding.

File: handsontable/src/editors/dateEditor/dateEditor.js (hunk 4 above)
Comment (jansiegel) [reply handsontable/src/editors/dateEditor/dateEditor.js:88]: Honestly, I tried adding a test for that but failed 🙄 Namely, the caret position seemed to be stuck at the same position, regardless of any keypress simulations I did. Do you maybe have any tips on how to potentially overcome this?

Comment (budnix) [reply handsontable/src/editors/dateEditor/dateEditor.js:88]: I see, now. I've tried to prepare some code but without success. The `event.code` property is always `undefined` using simulated events. I'd leave it for feature improvements.

**Review by budnix — APPROVED** [review]

**Comment by AMBudnik** [comment]
> I don't know if it's out of the scope of this PR, but it's impossible to navigate through the Handsontable cell-type editor. Only vertical navigation works. Kapture 2023-10-13 at 10 07 37 [ Kapture 2023-10-13 at 10 07 37 ](https://private-user-images.githubusercontent.com/571316/274838079-6bf2eb83-b294-4dbc-87bb-3cf6dc79f283.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE3NzI0NjYsIm5iZiI6MTcwMTc3MjE2NiwicGF0aCI6Ii81NzEzMTYvMjc0ODM4MDc5LTZiZjJlYjgzLWIyOTQtNGRiYy04N2JiLTNjZjZkYzc5ZjI4My5naWY_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBSVdOSllBWDRDU1ZFSDUzQSUyRjIwMjMxMjA1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDIzMTIwNVQxMDI5MjZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zZDcyMjBkYjBjZWJlOTFiYjM5NmNhNzg4M2Q2ZTMwZTk5MzMzNGU1MTJiYWNlZWVmNmI5MThhNTFiMmUzZmYxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.nPfp1z5gz8apefuh4-XHBP_7U-WLxVE2aocm-nywAdA) [ ](https://private-user-images.githubusercontent.com/571316/274838079-6bf2eb83-b294-4dbc-87bb-3cf6dc79f283.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE3NzI0NjYsIm5iZiI6MTcwMTc3MjE2NiwicGF0aCI6Ii81NzEzMTYvMjc0ODM4MDc5LTZiZjJlYjgzLWIyOTQtNGRiYy04N2JiLTNjZjZkYzc5ZjI4My5naWY_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBSVdOSllBWDRDU1ZFSDUzQSUyRjIwMjMxMjA1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDIzMTIwNVQxMDI5MjZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zZDcyMjBkYjBjZWJlOTFiYjM5NmNhNzg4M2Q2ZTMwZTk5MzMzNGU1MTJiYWNlZWVmNmI5MThhNTFiMmUzZmYxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.nPfp1z5gz8apefuh4-XHBP_7U-WLxVE2aocm-nywAdA)

 reported at https://github.com/handsontable/dev-handsontable/issues/1636

**Pull request merged**